### PR TITLE
(Ready for Review) Integration Tests for SDK

### PR DESCRIPTION
**Pull request by dbarbato-sf**

### CODE REVIEW

**Comment by dbarbato-sf** [comment]
Need to update circleCi config to run integration tests.

File: integrationTests/package.json [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    "lint": "eslint \"**/*.ts\" --quiet --fix",
+    "pretest": "npm run lint && npm run depcheck",
+    "test": "nyc mocha \"tests/*.test.*\"",
+    "fetch:internal": "npm pack ../packages/core && npm pack ../packages/generator && npm install ",
+    "test:integration": "npm run fetch:internal &&  npm run test",
+    "test:integration:redis": "fetch:internal && npm run redis --prefix ../packages/core && npm run test; npm run redis:shutdown --prefix ../packages/core",
+    "test:integration:redis:debug": "fetch:internal && npm run redis --prefix ../packages/core && npm run test; npm run redis:shutdown:debug --prefix ../packages/core"
+  },
+  "dependencies": {
+    "@commerce-apps/core": "file:./commerce-apps-core-1.4.1-beta.0.tgz",
```
Comment (monkey-jeff) [inline integrationTests/package.json:17]: How do we keep this version in line when we update the version?

File: integrationTests/package.json [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  "name": "commerce-sdk-integration-test",
+  "version": "1.4.1-beta.0",
+  "description": "Salesforce Commerce Cloud SDK Integration Test",
+  "homepage": "https://developer.commercecloud.com/s/",
+  "scripts": {
+    "depcheck": "depcheck",
+    "lint": "eslint \"**/*.ts\" --quiet --fix",
+    "pretest": "npm run lint && npm run depcheck",
+    "test": "nyc mocha \"tests/*.test.*\"",
+    "fetch:internal": "npm pack ../packages/core && npm pack ../packages/generator && npm install ",
```
Comment (monkey-jeff) [inline integrationTests/package.json:11]: Does this drop the packages into the current working directory?

Comment (monkey-jeff) [reply integrationTests/package.json:11]: Also why fetch?

File: integrationTests/package.json [hunk 3]
```diff
@@ -0,0 +1,95 @@
+{
+  "name": "commerce-sdk-integration-test",
```
Comment (monkey-jeff) [inline integrationTests/package.json:2]: Should add internal: false here as this package is never published.

File: integrationTests/package.json [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    "cache": false,
+    "include": [
+      "node_modules/@commerce-apps/core/dist/**"
+    ],
+    "excludeNodeModules": false,
+    "source-map": false,
+    "per-file": true,
+    "reporter": [
+      "lcov",
+      "text"
```
Comment (monkey-jeff) [inline integrationTests/package.json:31]: I think we should drop "text" from here and just have the lcov output for us.

File: integrationTests/package.json [hunk 5]
```diff
@@ -0,0 +1,95 @@
+{
```
Comment (monkey-jeff) [inline integrationTests/package.json:1]: How does this get executed by circleci?

File: integrationTests/package.json [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  "name": "commerce-sdk-integration-test",
+  "version": "1.4.1-beta.0",
+  "description": "Salesforce Commerce Cloud SDK Integration Test",
+  "homepage": "https://developer.commercecloud.com/s/",
+  "scripts": {
+    "depcheck": "depcheck",
+    "lint": "eslint \"**/*.ts\" --quiet --fix",
```
Comment (monkey-jeff) [inline integrationTests/package.json:8]: We are linting for only ts files here?

File: integrationTests/package.json [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  },
+  "dependencies": {
+    "@commerce-apps/core": "file:./commerce-apps-core-1.4.1-beta.0.tgz",
+    "commerce-sdk": "file:./commerce-sdk-1.4.1-beta.0.tgz"
+  },
+  "nyc": {
+    "all": true,
+    "cache": false,
+    "include": [
+      "node_modules/@commerce-apps/core/dist/**"
```
Comment (monkey-jeff) [inline integrationTests/package.json:24]: this is only including the core package right now.

File: integrationTests/tsconfig.json [hunk 8]
```diff
@@ -0,0 +1,15 @@
+{
+    "extends": "../base.tsconfig.json",
+    // NOTE: Everything except what you want compiled into JS should be in the exclusions.
+    "compilerOptions": {
+        "outDir": "./dist"
+    },
+    "exclude": [
+        "node_modules",
+        "gulpfiles",
```
Comment (monkey-jeff) [inline integrationTests/tsconfig.json:9]: not needed here as we don't have any gulp files

File: packages/core/package.json [hunk 9]
```diff
@@ -8,9 +8,10 @@
     "depcheck": "depcheck",
     "pretest": "npm run lint && npm run depcheck",
     "redis": "docker rm --force redis-commerce-sdk; docker run -dp 6379:6379 --name redis-commerce-sdk redis:5-alpine",
-    "redis:shutdown": "docker rm --force redis-commerce-sdk; docker rm --force redis-commerce-sdk-debug",
+    "redis:shutdown": "docker rm --force redis-commerce-sdk",
     "redis:debug": "docker rm --force redis-commerce-sdk-debug; docker run -dp 6379:6379 --name redis-commerce-sdk-debug redis:5-alpine && docker exec -it redis-commerce-sdk-debug redis-cli monitor",
-    "test": "nyc mocha \"test/**/*.test.ts\" --reporter=xunit --reporter-options output=../../reports/core.xml",
+    "redis:debug:shutdown": "docker rm --force redis-commerce-sdk-debug",
+    "test": "nyc mocha \"test/**/*.test.ts\" --reporter=xunit --reporter-options output=../../reports_integration/core.xml",
```
Comment (monkey-jeff) [inline packages/core/package.json:14]: Why was this moved?

File: integrationTests/package.json (hunk 5 above)
Comment (dbarbato-sf) [reply integrationTests/package.json:1]: I noted yesterday that this needs to be addressed, will talk with @kalagarraj to enable this.

File: integrationTests/package.json (hunk 3 above)
Comment (dbarbato-sf) [reply integrationTests/package.json:2]: @monkey-jeff: internal: false, or internal: true? internal: false would imply it's not internal and should be published.

File: integrationTests/package.json (hunk 2 above)
Comment (dbarbato-sf) [reply integrationTests/package.json:11]: > Does this drop the packages into the current working directory?
@monkey-jeff Yes it does, is that an issue?

Comment (dbarbato-sf) [reply integrationTests/package.json:11]: > Also why fetch?

@monkey-jeff because it's fetching the SDK and pulling it into the integration test package...

File: integrationTests/package.json (hunk 1 above)
Comment (dbarbato-sf) [reply integrationTests/package.json:17]: I'll test if you can use a wildcard, If not, I didn't see a way to re-use the version from the package.json as a variable, but I'll take another look

File: integrationTests/package.json (hunk 7 above)
Comment (dbarbato-sf) [reply integrationTests/package.json:24]: I thought a while ago we said we only wanted to support core, but I'll add generator now.

File: integrationTests/tsconfig.json (hunk 8 above)
Comment (dbarbato-sf) [reply integrationTests/tsconfig.json:9]: done

File: integrationTests/package.json (hunk 4 above)
Comment (dbarbato-sf) [reply integrationTests/package.json:31]: done

File: packages/core/package.json (hunk 9 above)
Comment (dbarbato-sf) [reply packages/core/package.json:14]: It wasn't moved, a new line was inserted to support redis debug shutdown and the name of the target report folder was also updated (as it's not integration tests). Which makes it look like the line moved.

File: integrationTests/package.json (hunk 6 above)
Comment (dbarbato-sf) [reply integrationTests/package.json:8]: Fixed

File: integrationTests/package.json (hunk 2 above)
Comment (monkey-jeff) [reply integrationTests/package.json:11]: I would rename it since in javascript land fetch is the primary used http request library.  Where the "fetch api" has become a standard.

File: integrationTests/package.json (hunk 3 above)
Comment (monkey-jeff) [reply integrationTests/package.json:2]: that is correct..wrote this pre-coffee

File: packages/core/package.json (hunk 9 above)
Comment (monkey-jeff) [reply packages/core/package.json:14]: Sorry what i mean is 
```
    "test": "nyc mocha \"test/**/*.test.ts\" --reporter=xunit --reporter-options output=../../reports/core.xml",
```
became
```
    "test": "nyc mocha \"test/**/*.test.ts\" --reporter=xunit --reporter-options output=../../reports_integration/core.xml",
```

Which moved the output of the report.

File: integrationTests/package.json (hunk 2 above)
Comment (kalagarraj) [reply integrationTests/package.json:11]: how about pack:sdk?

File: integrationTests/package.json (hunk 5 above)
Comment (kalagarraj) [reply integrationTests/package.json:1]: It is enabled in my PR, you can merge after it is approved

File: integrationTests/package.json (hunk 2 above)
Comment (dbarbato-sf) [reply integrationTests/package.json:11]: renamed to bundle:sdk

File: packages/core/package.json (hunk 9 above)
Comment (dbarbato-sf) [reply packages/core/package.json:14]: because this is no longer an integration test..

File: integrationTests/package.json (hunk 1 above)
Comment (dbarbato-sf) [reply integrationTests/package.json:17]: resolved

File: packages/core/package.json (hunk 9 above)
Comment (dbarbato-sf) [reply packages/core/package.json:14]: Yep, sorry, this was left over from when we were going to put integration tests in code. Resolved it now.

File: integrationTests/package.json [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+  "name": "commerce-sdk-integration-test",
+  "version": "1.4.1-beta.0",
+  "description": "Salesforce Commerce Cloud SDK Integration Test",
+  "homepage": "https://developer.commercecloud.com/s/",
+  "private": true,
+  "scripts": {
+    "depcheck": "depcheck",
+    "lint": "eslint . --ext .ts,.js --quiet --fix",
+    "pretest": "npm run lint && npm run depcheck",
+    "test": "nyc mocha \"tests/*.test.*\"",
```
Comment (wjhsf) [inline integrationTests/package.json:11]: Would this try to run `.test.helper.ts` files? You may want to change this to specifically `.test.ts` and `.test.js`

File: integrationTests/tsconfig.json [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+    // NOTE: Everything except what you want compiled into JS should be in the exclusions.
+    "compilerOptions": {
+        "outDir": "./dist"
+    },
+    "exclude": [
+        "node_modules",
+        "dist"
+    ],
+    "include": [
+        "./renderedTemplates"
```
Comment (wjhsf) [inline integrationTests/tsconfig.json:12]: This is specific to the core package, it's not needed here. And I don't think compilerOptions and exclude are needed either, because nothing is ever compiled here.

File: integrationTests/package.json (hunk 10 above)
Comment (dbarbato-sf) [reply integrationTests/package.json:11]: I didn't know that .test.helper.ts was a thing.

File: integrationTests/tsconfig.json (hunk 11 above)
Comment (dbarbato-sf) [reply integrationTests/tsconfig.json:12]: Aren't the .ts files transpiled and use these options?

File: commerce-sdk.code-workspace [hunk 12]
```diff
@@ -16,6 +16,10 @@
 		{
 			"name": "Generator",
 			"path": "packages/generator"
+		},
+		{
+			"name": "IntegrationTests",
```
Comment (wjhsf) [inline commerce-sdk.code-workspace:21]: The name is for humans. Add a space!

File: integrationTests/package.json [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+    "bundle:sdk": "npm pack ../packages/generator && mv commerce-sdk-*.tgz commerce-sdk.tgz",
+    "bundle:all": "npm run bundle:core && npm run bundle:sdk",
+    "bundle:install": "npm run bundle:all && npm install",
+    "bundle:clean:install": "npm run clean:all && npm install",
+    "test:integration": "npm run test",
+    "test:integration:redis": "npm run redis && npm run test; npm run redis:shutdown",
+    "test:integration:redis:debug": "npm run redis:debug && npm run test; npm run redis:debug:shutdown",
+    "test:integration:bundle": "npm run bundle:install &&  npm run test",
+    "test:integration:bundle:redis": "bundle:install && npm run redis && npm run test; npm run redis:shutdown",
+    "test:integration:bundle:redis:debug": "bundle:install && npm run redis:debug && npm run test; npm run redis:debug:shutdown",
```
Comment (wjhsf) [inline integrationTests/package.json:28]: `bundle:install` should be `npm run bundle:install`.

File: integrationTests/tests/authHelper.test.js [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+"use strict";
+const chai = require("chai");
+const { stripBearer } = require("@commerce-apps/core");
+
+describe("JS: Auth Helper Strip 0Bearer", () => {
```
Comment (wjhsf) [inline integrationTests/tests/authHelper.test.js:11]: typo?

File: package.json [hunk 15]
```diff
@@ -4,10 +4,27 @@
   "scripts": {
     "postinstall": "lerna bootstrap --hoist",
     "updateApis": "lerna run updateApis",
+
     "build": "lerna run build",
+    "redis": "npm run redis --prefix ./core",
+    "redis:shutdown": "npm run redis:shutdown --prefix ./core",
+    "redis:debug": "npm run redis:debug --prefix ./core",
+    "redis:debug:shutdown": "npm run redis:debug:shutdown --prefix ./core",
```
Comment (wjhsf) [inline package.json:12]: Shouldn't `prefix` be `./packages/core`?

File: package.json [hunk 16]
```diff
@@ -4,10 +4,27 @@
...
+    "test:integration:bundle": "lerna run test:integration",
+    "test:integration:bundle:redis": "lerna run test:integration:redis",
+    "run test:integration:bundle:redis:debug": "lerna run test:integration:redis:debug",
+    "test:integration": "npm run test:integration:bundle --prefix ./integrationTests",
+    "test:integration:redis": "npm run test:integration:redis --prefix ./integrationTests",
+    "run test:integration:redis:debug": "npm run test:integration:redis:debug --prefix ./integrationTests",
+    "test:all": "npm run test && npm run test:integration",
+    "test:all:redis": "npm run test && npm run test:integration:bundle:redis",
+    "test:all:redis:debug": "npm run test && npm run test:integration:bundle:redis:debug",
+    "npm run clean": "lerna run clean",
```
Comment (wjhsf) [inline package.json:24]: Do we also want to add a clean script for the integration tests here?

File: integrationTests/tests/authHelper.test.ts [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+"use strict";
+import chai, { expect } from "chai";
+import { stripBearer } from "@commerce-apps/core";
+
+describe("TS: Auth Helper Strip Bearer", () => {
```
Comment (wjhsf) [inline integrationTests/tests/authHelper.test.ts:11]: Is there any better way to differentiate JS/TS tests? (There might not be...)

File: package.json [hunk 18]
```diff
@@ -4,10 +4,27 @@
...
+    "redis": "npm run redis --prefix ./core",
+    "redis:shutdown": "npm run redis:shutdown --prefix ./core",
+    "redis:debug": "npm run redis:debug --prefix ./core",
+    "redis:debug:shutdown": "npm run redis:debug:shutdown --prefix ./core",
     "test": "lerna run test",
     "test:debug": "lerna run test:debug",
-    "test:integration": "lerna run test:integration",
+    "test:integration:bundle": "lerna run test:integration",
+    "test:integration:bundle:redis": "lerna run test:integration:redis",
+    "run test:integration:bundle:redis:debug": "lerna run test:integration:redis:debug",
```
Comment (kalagarraj) [inline package.json:17]: No need to add "run" prefix for custom scripts.

File: package.json [hunk 19]
```diff
@@ -4,10 +4,27 @@
...
+    "test:integration": "npm run test:integration:bundle --prefix ./integrationTests",
+    "test:integration:redis": "npm run test:integration:redis --prefix ./integrationTests",
+    "run test:integration:redis:debug": "npm run test:integration:redis:debug --prefix ./integrationTests",
+    "test:all": "npm run test && npm run test:integration",
+    "test:all:redis": "npm run test && npm run test:integration:bundle:redis",
+    "test:all:redis:debug": "npm run test && npm run test:integration:bundle:redis:debug",
```
Comment (kalagarraj) [inline package.json:23]: Should this be "npm run clean" just clean?

File: package.json [hunk 20]
```diff
@@ -4,10 +4,27 @@
...
+    "npm run clean": "lerna run clean",
+    "npm run clean:node": "lerna run clean:node",
```
Comment (kalagarraj) [inline package.json:25]: same comment as above, rename to clean:node

File: packages/generator/package.json [hunk 21]
```diff
@@ -18,16 +18,19 @@
   ],
   "scripts": {
     "depcheck": "depcheck",
+    "lint": "eslint . --ext .ts --quiet --fix",
     "pretest": "npm run lint && npm run depcheck",
-    "test:debug": "nyc mocha \"test/**/*.test.ts\"",
+
+    "build": "npm run renderTemplates && tsc && eslint dist --ext .ts,.js --quiet --fix",
     "test": "nyc mocha \"test/**/*.test.ts\" --reporter=xunit --reporter-options output=../../reports/generator.xml",
-    "test:integration": "nyc --check-coverage=false mocha \"test_integration/**/*.ts\" --retries 2 --timeout 5000 --reporter=xunit --reporter-options output=../../reports/generator-integration.xml",
```
Comment (kalagarraj) [inline packages/generator/package.json:24]: Is there a reason removing test:integration?

File: integrationTests/package.json [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+    "lint": "eslint . --ext .ts,.js --quiet --fix",
+    "pretest": "npm run lint && npm run depcheck",
+    "test:ts": "nyc mocha \"tests/*.test.ts\"",
+    "test:js": "nyc mocha \"tests/*.test.js\"",
+    "test": "npm run test:ts && npm run test:js",
+    "redis": "npm run redis --prefix ../packages/core",
+    "redis:shutdown": "npm run redis:shutdown --prefix ../packages/core",
+    "redis:debug": "npm run redis:debug --prefix ../packages/core",
+    "redis:debug:shutdown": "npm run redis:debug:shutdown --prefix ../packages/core",
+    "bundle:core": "npm pack ../packages/core && mv commerce-apps-core*.tgz commerce-apps-core.tgz",
```
Comment (kalagarraj) [inline integrationTests/package.json:18]: Potential one off issue for bundle:core script...
Can u pls test the following and see if it works..

1. npm run bundle:core
2. Change the version in core package
3. npm run bundle:core

File: integrationTests/package.json [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+    "bundle:sdk": "npm pack ../packages/generator && mv commerce-sdk-*.tgz commerce-sdk.tgz",
+    "bundle:all": "npm run bundle:core && npm run bundle:sdk",
+    "bundle:install": "npm run bundle:all && npm install",
+    "bundle:clean:install": "npm run clean:all && npm install",
```
Comment (kalagarraj) [inline integrationTests/package.json:22]: I think this script will fail every time. Can u please test this bundle:clean:install? if it is not used anywhere else pls remove it..

File: integrationTests/package.json [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+    "test:integration": "npm run test",
+    "test:integration:redis": "npm run redis && npm run test; npm run redis:shutdown",
+    "test:integration:redis:debug": "npm run redis:debug && npm run test; npm run redis:debug:shutdown",
+    "test:integration:bundle": "npm run bundle:install &&  npm run test",
```
Comment (kalagarraj) [inline integrationTests/package.json:26]: u may need to call clean before bundle install

File: package.json (hunk 18 above)
Comment (dbarbato-sf) [reply package.json:17]: good catch, copy paste error.

File: packages/generator/package.json (hunk 21 above)
Comment (dbarbato-sf) [reply packages/generator/package.json:24]: Yes, integration tests should now all live in the integration test folder.

Comment (kalagarraj) [reply packages/generator/package.json:24]: These tests are not part of integrationTests folder yet. Until it is moved, we should keep testing those

File: integrationTests/package.json (hunk 23 above)
Comment (dbarbato-sf) [reply integrationTests/package.json:22]: Which bit do you think will fail? I saw an issue with the rm part of the clean failing when there was no file, I've added a fix for that.

File: packages/generator/package.json (hunk 21 above)
Comment (dbarbato-sf) [reply packages/generator/package.json:24]: For some reason I was thinking your story moved those.

Comment (dbarbato-sf) [reply packages/generator/package.json:24]: Talking with @monkey-jeff we'll leave these disabled for now as they don't provide huge value and re-address the tests in W-7487432.

File: integrationTests/tests/authHelper.test.ts (hunk 17 above)
Comment (dbarbato-sf) [reply integrationTests/tests/authHelper.test.ts:11]: Open to suggestions.

File: integrationTests/tsconfig.json [hunk 25]
```diff
@@ -0,0 +1,11 @@
+{
+    "extends": "../base.tsconfig.json",
+    // NOTE: Everything except what you want compiled into JS should be in the exclusions.
+    "compilerOptions": {
+        "outDir": "./dist"
```
Comment (monkey-jeff) [inline integrationTests/tsconfig.json:5]: Are we transpiling the tests?

File: package.json [hunk 26]
```diff
@@ -5,9 +5,24 @@
...
     "test:debug": "lerna run test:debug",
-    "test:integration": "lerna run test:integration",
+    "test:integration:bundle": "lerna run test:integration",
+    "test:integration:bundle:redis": "lerna run test:integration:redis",
+    "test:integration:bundle:redis:debug": "lerna run test:integration:redis:debug",
+    "test:integration": "npm run test:integration:bundle --prefix ./integrationTests",
+    "test:integration:redis": "npm run test:integration:redis --prefix ./integrationTests",
+    "run test:integration:redis:debug": "npm run test:integration:redis:debug --prefix ./integrationTests",
+    "test:all": "npm run test && npm run test:integration",
+    "test:all:redis": "npm run test && npm run test:integration:bundle:redis",
```
Comment (kalagarraj) [inline package.json:21]: Add --prefix integrationTests

File: package.json [hunk 27]
```diff
@@ -5,9 +5,24 @@
...
+    "redis:debug:shutdown": "npm run redis:debug:shutdown --prefix ./package/core",
     "test": "lerna run test",
     "test:debug": "lerna run test:debug",
-    "test:integration": "lerna run test:integration",
+    "test:integration:bundle": "lerna run test:integration",
+    "test:integration:bundle:redis": "lerna run test:integration:redis",
+    "test:integration:bundle:redis:debug": "lerna run test:integration:redis:debug",
+    "test:integration": "npm run test:integration:bundle --prefix ./integrationTests",
+    "test:integration:redis": "npm run test:integration:redis --prefix ./integrationTests",
+    "run test:integration:redis:debug": "npm run test:integration:redis:debug --prefix ./integrationTests",
```
Comment (kalagarraj) [inline package.json:19]: remove run

File: package.json (hunk 26 above)
Comment (dbarbato-sf) [reply package.json:21]: no, this is an internal call to the script on line 15

File: package.json (hunk 27 above)
Comment (dbarbato-sf) [reply package.json:19]: Damn, I thought I got all these.
